Add `exact` Param To ObjectTypeAnnotations

wmonk opened the pull request:
<!--
Before making a PR please make sure to read our contributing guidelines
https://github.com/babel/babel/blob/master/CONTRIBUTING.md

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. It should be underlined in the preview if done correctly.
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | 👍 
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | 
| Documentation PR         | <!-- If so, add `[skip ci]` to your commit message to skip CI -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->

Currently if you try to use `t.objectTypeAnnotation` it fails, because while there is documentation for the param `exact` it is not added in the `visitor` entry. This PR adds `exact` and an optionally validated param for `ObjectTypeAnnotation`.

I'm not sure if removing this was intentional, but the documentation is quite confusing if it was. I think it was probably just a mistake.

nicolo-ribaudo commented on packages/babel-types/src/definitions/flow.js:
```diff
@@ -250,15 +250,15 @@ defineType("NumberTypeAnnotation", {
 
 defineType("ObjectTypeAnnotation", {
-  visitor: ["properties", "indexers", "callProperties"],
+  visitor: ["properties", "indexers", "callProperties", "exact"],
```
`visitors` should only contain `Node` properties, since it represents the properties which should be visited by babel-traverse.
What you need is `builder`, which represents the parameters passed to the `t.objectTypeAnnotation` function.

nicolo-ribaudo commented on packages/babel-types/src/definitions/flow.js:
```diff
@@ -250,15 +250,15 @@ defineType("NumberTypeAnnotation", {
     indexers: validateOptional(arrayOfType("ObjectTypeIndexer")),
     callProperties: validateOptional(arrayOfType("ObjectTypeCallProperty")),
-    exact: validate(assertValueType("boolean")),
+    exact: validateOptional(assertValueType("boolean")),
```
Why is this optional? It should be required but defaulting to `false` (I don't remember if it is actually possible to set default properties value)

wmonk commented:
I've replaced this with builder, which i'm not sure is actually right when this get used in the `@babel/generator` module. Would be good to get some guidance 😃 

I've added a test now in `@babel/generator` that shows the correct outputted exact type.

nicolo-ribaudo commented:
You can see what I meant by "builder" here: https://github.com/wmonk/babel/blob/4fc40bdc21b96265c724daa50bb072316739ce58/packages/babel-types/src/definitions/core.js#L48-L49.

babel-generator was correct, you should revert its changes.

nicolo-ribaudo commented on packages/babel-types/src/definitions/flow.js:
```diff
@@ -250,15 +251,16 @@ defineType("NumberTypeAnnotation", {
     indexers: validateOptional(arrayOfType("ObjectTypeIndexer")),
     callProperties: validateOptional(arrayOfType("ObjectTypeCallProperty")),
-    exact: validate(assertValueType("boolean")),
+    exact: validateDefault(assertValueType("boolean"), false),
```
I'd prefer this to be written as an object instead of using the `validateDefault` function, since `validator` and `default` have very little in common.

wmonk commented:
@nicolo-ribaudo thanks for pointing me in the right direction, makes sense now I think! The latest change should be correct now. I added a test to `babel/generator` anyway - think that's right?

wmonk commented:
😄 thanks!
